index.bs

```diff
@@ -5301,27 +5301,28 @@ partial interface MLGraphBuilder {
 <div>
     **Arguments:**
         - *input*: an {{MLOperand}}. The input 2-D tensor.
+        - *axis*: an {{unsigned long}} scalar. The dimension the softmax will be performed on.
 
     **Returns:**
         - an {{MLOperand}}. The output 2-D tensor that contains the softmax results, of the same shape as *input*.
 </div>
 
 <details open algorithm>
   <summary>
-    The <dfn method for=MLGraphBuilder>softmax(|input|)</dfn> method steps are:
+    The <dfn method for=MLGraphBuilder>softmax(|input|, |axis|)</dfn> method steps are:
   </summary>
     1. If [=MLGraphBuilder/validating operand=] with [=this=] and |input| returns false, then [=exception/throw=] a {{TypeError}}.
     1. If |input|'s [=MLOperand/rank=] is not 2, then [=exception/throw=] a {{TypeError}}.
```

Comment:
The `axis` needs to be validated according to input's rank.

Reply:
Added for the operator version in 99e87734fa8daa7f69c95d18c8c075fc6ed0b592

For the activation version - would that be done synchronously when the activation is provided in a graph builder call, or at build time? Do we have other cases of activation validation?

Reply:
All the information should be available at construction time (pre-`build`), since WebNN requires explicit shapes during construction.

> Do we have other cases of activation validation?

None come to mind, except possibly [softplus steepness](https://github.com/webmachinelearning/webnn/issues/645), but that's not impacted by the tensor shape constraints.

Reply:
> > For the activation version - would that be done synchronously when the activation is provided in a graph builder call,
> 
> All the information should be available at construction time (pre-`build`), since WebNN requires explicit shapes during construction.

Agreed. In particular, I suppose the validation should be done at the build method of operators that accept a fused activation function, like conv2d.

Reply:
Here's a sketch of what this could look like:

* In [create an `MLActivation`](https://webmachinelearning.github.io/webnn/#create-an-mlactivation), add optional _validation steps_ parameter, store the validation steps in an internal slot. The default validation steps are to return true.
* When creating softmax activation, pass these steps as the _validation steps_:
  1. If _axis_ is greater than or equal to _input_’s rank, then return false. 
  1. Otherwise, return true.
  * Note that if this were C++, this would be a lambda function, capturing _axis_. 
* In all 7 methods that take activation functions (singular or multiple), add steps like this:
  1. If _options_.activation exists, and running its _validation steps_ with _input_ returns false, then throw a `TypeError`.

 Thoughts:
* This could also be bundled into the [validate activation](https://webmachinelearning.github.io/webnn/#mlgraphbuilder-validate-activation) steps - make those take an `MLOperand` in addition to or instead of `MLGraphBuilder`.
* Is passing a single `MLOperand` (i.e. _input_) sufficient for validation?

WDYT?

Reply:
> * Is passing a single `MLOperand` (i.e. _input_) sufficient for validation?

I suppose it should pass _output_ MLOperand since the activation follows the operator that fuses it.

Reply:
I guess we'll need to pass the output descriptor.

Reply:
> In particular, I suppose the validation should be done at the build method of operators that accept a fused activation function, like conv2d.

Yep, since stand-alone activations adopt whatever the size is of the operator they are joined to or used inside. So full validation must be deferred for these.

```
MLConv2dOptions::activation
MLConvTranspose2dOptions::activation
MLBatchNormalizationOptions::activation
MLGruOptions::activations
MLGruCellOptions::activations
MLLstmOptions::activations
MLLstmCellOptions::activations
```

> Is passing a single MLOperand (i.e. input) sufficient for validation?

Sounds fine, the MLActivation and at least either the MLOperand.shape or the MLOperand itself for more generality.

Reply:
Okay - 7981fa52a0cdceba56c5026ab2e1021451d35306 sketches out the validation. Notes:

* I went with passing an `MLOperandDescriptor`, because earlier we decided that we'd do all validation _before_ creating the operator and output `MLOperand` to pass. 
* The phrasing around creating/passing a lambda is based on [Fetch](https://fetch.spec.whatwg.org/#body-fully-read)
* The GRU/LSTM outputs are complicated. I'm not sure I'm passing the right thing, so please please double check.